Fixes issue 816 (space in DS names) and add tenant name validation during tenant create

lipingxue opened the pull request:
This PR includes:
1. Fix issue 816(space in DS names)
2. validate tenant name during tenant create, tenant name are only allowed to be [a-zA-Z0-9_][a-zA-Z0-9_.-]*

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -854,11 +854,14 @@ def findDeviceByPath(vmdk_path, vm):
+        # Assume there is one space between these two parts "[<datastore name>]" and
+        # <parent-directory>/tenant/<vmdk-descriptor-name>
+        logging.debug("findDeviceByPath: d.backing.fileName %s", d.backing.fileName)
+        backing_disk = d.backing.fileName.rsplit(" ", 1)[1]
```
Can  `<parent-directory>/tenant/<vmdk-descriptor-name>` hasve a space ? The current code is guaranteed to fail. Maybe it's better to look for ']' ? Something like
```python
    ds, file  = d.backing.fileName.rsplit("]", 1)[1]
    datastore = ds[1:]
    backing_disk = file[1:]
```
That raises a couple of additional things to check:
(1) can we create a tenant with a space 
(2) can we create a volume with a space in the name.

shaominchen replied:
IMO, we shouldn't allow spaces in tenant & volume names. It should be OK to have spaces in tenant description though.

lipingxue replied:
I don't know space in "volume names" is allowed in vsphere or not? "datastore" names can have space.

msterin replied:
That was the point - please check it :-). Directory names (i.e. tenant) and volume names (i.e. files) will accept spaces unless we block them on creation. So please check if we do block it on creation.  I agree with @shaominchen that it's better to block spaces in tenant names and volume names. on creation.
With that said, i still think the code I suggested is slightly more resilient than looking for spaces.

lipingxue replied:
Yes. I agree with you.  I will modify the code as you suggested. Currently, I think we don't check for tenant name and volume name during creation. I will add the check logic.

lipingxue replied:
I did some test.
Vsphere allows volume name including space since I can create a VM with name including space.

But if I create a volume using local driver, it does not allow volume name to include space, and I got the following error.

lipingxue replied:
Added code to check tenant name and volume name during creation.

shaominchen commented on esx_service/vmdk_ops.py:
```diff
@@ -854,11 +854,14 @@ def findDeviceByPath(vmdk_path, vm):
         #   "[<datastore name>] <parent-directory>/tenant/<vmdk-descriptor-name>"
-        logging.debug("d.backing.fileName %s", d.backing.fileName)
-        backing_disk = d.backing.fileName.split(" ")[1]
+        # Assume there is one space between these two parts "[<datastore name>]" and
```
This comment is not very clear to me. I'd suggest to give an example in the comment so that it's easier to understand the logic of ....rsplit(" ", 1)[0]

lipingxue replied:
Done.

shaominchen commented on esx_service/vmdk_ops.py:
```diff
@@ -854,11 +854,14 @@ def findDeviceByPath(vmdk_path, vm):
+        logging.debug("findDeviceByPath: d.backing.fileName %s", d.backing.fileName)
+        backing_disk = d.backing.fileName.rsplit(" ", 1)[1]
 
         # datastore='[datastore name]'
```
Same as above: suggest to add an example in the comment to explain the logic below

lipingxue replied:
Done.

govint commented:
@lipingxue can you also include a test for the scenario in #892, if spaces are disallowed then that issue shouldn't be able to occur.

What about is a VSphere has already created a datastore with spaces in it. Then will that datastore be prevented from being used in any tenant. Ideally, user should be disallowed from adding that datastore.

govint commented on esx_service/utils/vmdk_utils.py:
```diff
@@ -232,7 +232,7 @@ def list_vmdks(path, volname="", show_snapshots=False):
     """
     Is the file a vmdk descriptor file?  We assume any file that ends in .vmdk,
-    does not have -delta or -flat or he likes at the end of filename, 
+    does not have -delta or -flat or he likes at the end of filename,
```
seems a typo, sentence must be corrected - "or he likes at the end of filename"

lipingxue replied:
@govint We only disallow space in tenant name and volume name during creation. So with this fix, issue #892 can still happen. We cannot control the datastore name. In VSphere, spaces in datastore name are allowed.

lipingxue replied:
Typo in comment is fixed too.

lipingxue commented:
Add code to validate tenant_name and volume name during tenant create and volume create.
1. create a tenant with invalid name
```
[root@localhost:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant create --name=.abc*d
Tenant name .abc*d is invalid, only [a-zA-Z0-9_][a-zA-Z0-9_.-]* is allowed

```

2. create a volume with invalid name
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name="abc*   edf"
Error response from daemon: create abc*   edf: VolumeDriver.Create: Volume name abc*   edf is invalid, only [a-zA-Z0-9][a-zA-Z0-9_.-]* is allowed
```

lipingxue commented:
@msterin @shaominchen @govint I have addressed your comments, change is repushed.  Please review it again.

msterin commented on esx_service/utils/vmdk_utils.py:
```diff
@@ -100,8 +100,11 @@ def get_datastores():
-        init_datastoreCache()
+    # if datastores == None:
+    #     init_datastoreCache()
+    # we should not use cached value, user may change datastore name on the fly
```
the problem is that init_datastoreCache is a an  expensive operation and it is called way too many times, often many times in the same call chain. The correct way to fix it would have been passing datastore list, but it was a very invasive change so the cache was introduced.  Let's discuss how to test and what are the other ways to fix the issue, rather than refetch all DS list every time

lipingxue replied:
As we discussed offline, will remove this change from this PR.

msterin commented on esx_service/utils/vmdk_utils.py:
```diff
@@ -100,8 +100,11 @@ def get_datastores():
-    if datastores == None:
-        init_datastoreCache()
+    # if datastores == None:
+    #     init_datastoreCache()
```
Why this change is introduced ? It seems to irrelevant to #816

lipingxue replied:
This change is needed. Think about this case:
step 1: assume we have datastore named "datastore1", which is the default_datastore, we create a volume "vol1"
step 2: rename "datastore1" to "datastore 1" (including space in the name)
step3: create another vol "vol2", which will fail.  

It will hit the error same as #818 . The root cause is "vmdk_utils.get_datastores" returns cached value which is outdated.

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -24,6 +24,10 @@
+import re
+
+# regex for valid tenant name
+VALID_TENANT_NAME_REG = "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
```
Need to add documentation for this, a one liner

shaominchen commented on esx_service/utils/vmdk_utils.py:
```diff
@@ -100,8 +100,11 @@ def get_datastores():
+    #     init_datastoreCache()
+    # we should not use cached value, user may change datastore name on the fly
+    # need to repull the info
+    init_datastoreCache()
```
With this change, we are losing the purpose of introducing datastore cache. Have we filed an issue to fix the datastore cache implementation?

lipingxue replied:
I discussed with Mark offline, we decided to remove this cache related change from this PR.

pdhamdhere replied:
Are you going to remove `init_datastoreCache()`

lipingxue replied:
No, we will keep init_datastoreCache().

msterin commented on esx_service/utils/auth_api.py:
```diff
@@ -239,9 +243,22 @@ def get_default_datastore(name):
 
+def is_valid_tenant_name(name):
+    """ Check the given name is a valid tenant name """
+    pattern = "^" + VALID_TENANT_NAME_REG + "$"
```
nit: why not pre-compile it once and keep the compiled regexp in a global ? It's not a bit deal since the operation is rare, but generally it's a waste.

shaominchen commented on esx_service/utils/auth_api.py:
```diff
@@ -239,9 +243,22 @@ def get_default_datastore(name):
     return error_info, default_datastore
 
+def is_valid_tenant_name(name):
+    """ Check the given name is a valid tenant name """
```
Nit: "Check..." => "Check if..."

shaominchen commented on esx_service/utils/auth_api.py:
```diff
@@ -239,9 +243,22 @@ def get_default_datastore(name):
+    """ Check the given name is a valid tenant name """
+    pattern = "^" + VALID_TENANT_NAME_REG + "$"
+    valid_name_reg = re.compile(pattern)
+    if valid_name_reg.match(name):
```
Nit: we can simplify it a little bit by removing the if-else block:

   return valid_name_reg.match(name)

msterin commented on esx_service/vmdk_ops.py:
```diff
@@ -153,12 +157,24 @@ def RunCommand(cmd):
 # for now we care about size and (maybe) policy
 def createVMDK(vmdk_path, vm_name, vol_name, opts={}, vm_uuid=None, tenant_uuid=None, datastore=None):
     logging.info("*** createVMDK: %s opts = %s", vmdk_path, opts)
+    if not is_valid_volume_name(vol_name):
```
I think we already have  code to parse volume names - please check if can re-use

lipingxue replied:
We have a util function parse_vol_name, which check whether the vol_name is match r"^.*-[0-9]{6}$" . So there is no limit on which character will be included in the vol_name. But docker local driver limit vol_name can only be "[a-zA-Z0-9][a-zA-Z0-9_.-]" . Do we want have the same limit as docker local driver does?

lipingxue replied:
@msterin

msterin replied:
parse_vol_name does more that that - it also separated DS name and handles them differnetly.
And no, I see no reason to mimic local local driver, we have deviated from it already by adding `@` to full volume name.

shaominchen commented on esx_service/vmdk_ops.py:
```diff
@@ -153,12 +157,24 @@ def RunCommand(cmd):
     return (s, o)
 
+def is_valid_volume_name(name):
+    """ Check the given name is a valid volume name """
```
Nit: "Check..." => "Check if..."

lipingxue replied:
Done. Remove this function and introduce a global variable "valid_volume_name_reg", use valid_volume_name_reg.match() to check directly.

shaominchen commented on esx_service/vmdk_ops.py:
```diff
@@ -153,12 +157,24 @@ def RunCommand(cmd):
+    """ Check the given name is a valid volume name """
+    pattern = "^" + VALID_VOLUME_NAME_REG + "$"
+    valid_name_reg = re.compile(pattern)
+    if valid_name_reg.match(name):
```
Nit: Same as above - we can simplify it a little bit:

   return valid_name_reg.match(name)

lipingxue replied:
Done.

pdhamdhere commented on esx_service/vmdk_ops.py:
```diff
@@ -129,6 +130,9 @@
 lockManager = threadutils.LockManager()
 
+# regex for valid volume name
+VALID_VOLUME_NAME_REG = "[a-zA-Z0-9][a-zA-Z0-9_.-]*"
```
Why do we need this one? We already validate volume name in `parse_vol_name`.

lipingxue replied:
docker local driver has this limitation on volume names during volume create. I think we should apply the same check.  We can discuss if you think differently. @msterin

msterin replied:
We do not care what local diver does, so unless there is a good reason for  the change allowed volnames, please remove this code. 

And, as Prashant has mentioned, the parsing is happening in `parse_vol_name` so any changes should be done there.

govint commented:
When trying out with a vol@<data store name with spaces> figured the below:

1. Run command to mkdir - failed because of spaces, quoting the path name
passed to mkdir worked.

2. find_devicebypath() in vmdk_ops.py has some code to get the datastore
name from a vmdk path and thats splitting the path incorrectly when the
datastore name has spaces in it.

Basically, fix an issue and then it breaks elsewhere. Keep fixing the
issues till the entire use case (create/get/remove/attach/detach) works
fine.

With (1) above fixed create is ok, attach/detach fail in (2) above.

On Fri, Feb 17, 2017 at 12:33 AM, Liping Xue <notifications@github.com>
wrote:

> *@lipingxue* commented on this pull request.
> ------------------------------
>
> In esx_service/vmdk_ops.py
> <https://github.com/vmware/docker-volume-vsphere/pull/905#discussion_r101600131>
> :
>
> > @@ -129,6 +130,9 @@
>  # For managing resource locks.
>  lockManager = threadutils.LockManager()
>
> +# regex for valid volume name
> +VALID_VOLUME_NAME_REG = "[a-zA-Z0-9][a-zA-Z0-9_.-]*"
>
> docker local driver has this limitation on volume names during volume
> create. I think we should apply the same check. We can discuss if you think
> differently. @msterin <https://github.com/msterin>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vmware/docker-volume-vsphere/pull/905#discussion_r101600131>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APHseFHwwZ83Xv3D12A0SN4_viiHON4Pks5rdJ2fgaJpZM4L9rmM>
> .
>

lipingxue commented:
@msterin I have addressed your comments:
1. remove "whitespace cleanup" from this PR
2. remove the volume name checking during volume creation

Please review it.

lipingxue commented:
Since Sam is on PTO, @msterin @govint @pdhamdhere please review it.

govint commented on esx_service/utils/auth_api.py:
```diff
@@ -241,7 +247,12 @@ def get_default_datastore(name):
+    if not valid_tenant_name_reg.match(name):
+        error_info = error_code.generate_error_info(ErrorCode.TENANT_NAME_INVALID, name, VALID_TENANT_NAME_REGEXP)
+        return error_info, None
+
```
Can this be a function like is_tenant_name_valid() vs. repeated in multiple locations.

lipingxue replied:
Done.

govint commented on esx_service/vmdk_ops_test.py:
```diff
@@ -715,20 +716,27 @@ def create_default_tenant_and_privileges(self):
+        if error_msg:
+            logging.warning(error_msg)
+            return
+
```
This action and error handling must be done before tenant create, its leaving a new tenant created and left in an unknown state (line 726 above).

lipingxue replied:
Addressed your comments. Now the logic is
1. check default_tenant exists or not
2. if default_tenant exists, do nothing, continue
3. if default_tenant does not exist, try to create, assert the create should not return error, since the test need to start with default_tenant exist

Similar logic also apply to default_privilege

shuklanirdesh82 commented on docs/user-guide/admin-cli.md:
```diff
@@ -39,17 +39,17 @@ optional arguments:
 A tenant named "_DEFAULT" will be created automatically post install.
 
-Creates a new named tenant and optionally assigns VMs.
+Creates a new named tenant and optionally assigns VMs. Valid tenant name is only allowed to be "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
```
@lipingxue  a quick question: PR description says ```Fixes issue 816 (space in DS names).``` can you please help to understand why some of the fixes are related with ```TENANT_NAME_INVALID``` or ```VOLUME_NAME_INVALID```?

lipingxue replied:
This tenant name validation fix is not related to issue 816 itself.  Sam pointed out earlier that in his comment to suggest that we should give some limitation to the tenant name during creation. For example, no space allowed in the tenant name, which will avoid some potential issues that may occur like issue 816 if user specify tenant name with spaces and other weird characters.

shuklanirdesh82 replied:
It would be helpful if you can add some useful information in the PR title as well as in the first comment. 

Such information will be useful for future reference while searching from title.  Current title proposing the ```Fixes issue 816 (space in DS names).```

lipingxue replied:
OK. Change the title.

shuklanirdesh82 commented on esx_service/vmdk_ops_test.py:
```diff
@@ -715,20 +716,27 @@ def create_default_tenant_and_privileges(self):
-        error_info, tenant_uuid, tenant_name = auth.get_default_tenant()
+        error_msg, tenant_uuid, tenant_name = auth.get_default_tenant()
+        if error_msg:
+            logging.warning(error_msg)
```
don't we need to fail the test here ( not able to create default tenant)?

lipingxue replied:
Yes. I addressed this. See above comments I replied to Govindan's comments.

shuklanirdesh82 commented on esx_service/vmdk_ops.py:
```diff
@@ -855,11 +856,10 @@ def findDeviceByPath(vmdk_path, vm):
-        # datastore='[datastore name]'
-        datastore = d.backing.fileName.split(" ")[0]
-        datastore = datastore[1:-1]
+        ds, disk_path = d.backing.fileName.rsplit("]", 1)
```
can you please confirm about having a test for this code snip?

lipingxue replied:
Yes. I added test for this.

shuklanirdesh82 commented on esx_service/utils/auth_api.py:
```diff
@@ -24,6 +24,12 @@
+import re
+
+# regex for valid tenant name
+VALID_TENANT_NAME_REGEXP = "[a-zA-Z0-9_][a-zA-Z0-9_.-]*"
```
1. Is it something we are introducing new?
2. What happens if customer is having tenant name that violates this rule? I guess new tenant creates enforces the restriction but would like to make sure here.

lipingxue replied:
Yes, the tenant create with invalid name should fail.

shuklanirdesh82 commented on esx_service/utils/auth_api.py:
```diff
@@ -241,7 +247,12 @@ def get_default_datastore(name):
     """ API to create a tenant """
+    global valid_tenant_name_reg
     logging.debug("_tenant_create: name=%s description=%s vm_list=%s privileges=%s", name, description, vm_list, privileges)
+    if not valid_tenant_name_reg.match(name):
```
1. need to add testing done for this code snip
2. can you confirm about having tests exercising this new change?

lipingxue commented:
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol2
vol2
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls                              
DRIVER              VOLUME NAME
vmdk                vol1@datastore    1
vmdk                vol2@datastore    1

lipingxue commented:
Test to make sure vol can be created on a datastore which name includes spaces.
Configuration: 
only _DEFAULT tenant is configured, VM "photon5" is created on "datastore1"
Step1: From VM"photon5", create a volume "vol1"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol1
vol1
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore1
```

Step2:
rename datastore "datastore1" to "datastore    1" (to insert spaces)
restart /etc/init.d/vmdk-opsd service. This is to avoid hitting the issue caused by stale datastore info returned by datastore cache(issue 818, which I am working on).  

Step3: create another volume "vol2"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore    1
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol2
vol2
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls                              
DRIVER              VOLUME NAME
vmdk                vol1@datastore    1
vmdk                vol2@datastore    1

```

lipingxue commented:
Test to make sure create tenant with invalid name will fail

create a tenant with invalid name like "tenant     abc" failed  as expect.
```
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4/dockvols] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant ls
Uuid                                  Name      Description               Default_datastore  VM_list  
------------------------------------  --------  ------------------------  -----------------  -------  
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default tenant                              
[root@localhost:/vmfs/volumes/57473702-8b550986-70c0-000c290799f4/dockvols] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant create --name="tenant     abc"
Tenant name tenant     abc is invalid, only [a-zA-Z0-9_][a-zA-Z0-9_.-]* is allowed

```

govint commented:
@lipingxue can you also add a test that creates the volume on a datastore with spaces in it like "docker volume create --driver=vmdk --name=vol1@data\ \ \ store"

lipingxue commented:
@govint Added the test you request
Configuration: only _DEFAULT tenant is configured, VM "photon5" is created on datastore1
Step1: create two volumes "vol1" and "vol2"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol1
vol1
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol2
vol2
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore1
vmdk                vol2@datastore1

```
Step2:
rename datastore "datastore1" to "datastore   1"
restart /etc/init.d/vmdk-opsd service. This is to avoid hitting the issue caused by stale datastore info returned by datastore cache(issue 818, which I am working on).

Step3: create a volume "vol3@datastore   1"
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name="vol3@datastore   1"
vol3@datastore   1

```

Step4: create a volume "vol4@datastore\ \ \ 1" with the following command
```
root@photon-eZ5ILvREQ [ ~ ]# docker volume create --driver=vmdk --name=vol4@datastore\ \ \ 1
vol4@datastore   1
root@photon-eZ5ILvREQ [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vmdk                vol1@datastore   1
vmdk                vol2@datastore   1
vmdk                vol3@datastore   1
vmdk                vol4@datastore   1

```

govint commented:
@lipingxue thanks and looks good.

shuklanirdesh82 commented on esx_service/vmdk_ops.py:
```diff
@@ -87,6 +87,7 @@
 import error_code
 import auth_api
 from error_code import ErrorCode
+import re
```
trivial: I guess this needs to be removed as well.

shuklanirdesh82 replied:
never mind .. not sure why diff is saying it is added as part of this change. Please ignore above comment. Sorry for the confusion.

Thanks!

lipingxue replied:
Done. Thanks for review.
